micrometer-core/src/main/java/io/micrometer/core/instrument/binder/mongodb/MongoMetricsCommandListener.java

```diff
@@ -15,59 +15,118 @@
+
+    public MongoMetricsCommandListener(MeterRegistry registry, int collectionNamesCacheMaxSize, int collectionNamesCacheOverflowErrorLogInterval) {
+        this(registry, collectionNamesCacheMaxSize, collectionNamesCacheOverflowErrorLogInterval < 1 ? null : new EveryNthOperation(collectionNamesCacheOverflowErrorLogInterval));
+    }
+
+    // VisibleForTesting
+    MongoMetricsCommandListener(MeterRegistry registry, int collectionNamesCacheMaxSize, @Nullable EveryNthOperation collectionNamesCacheOverflowErrorTracker) {
         this.registry = registry;
+        this.mongoCommandUtil = new MongoCommandUtil();
+        this.collectionNamesCache = new ConcurrentHashMap<>();
+        if (collectionNamesCacheMaxSize < 1) {
+            throw new IllegalArgumentException("collectionNamesCacheMaxSize must be set to a positive value");
+        }
+        this.collectionNamesCacheMaxSize = collectionNamesCacheMaxSize;
+        this.collectionNamesCacheOverflowErrorTracker = collectionNamesCacheOverflowErrorTracker;
     }
 
     @Override
-    public void commandStarted(CommandStartedEvent commandStartedEvent) {
+    public void commandStarted(CommandStartedEvent event) {
+        mongoCommandUtil.determineCollectionName(event.getCommandName(), event.getCommand())
+                .ifPresent(collectionName -> addCollectionNameToCache(event, collectionName));
     }
 
     @Override
     public void commandSucceeded(CommandSucceededEvent event) {
-        timeCommand(event, "SUCCESS", event.getElapsedTime(TimeUnit.NANOSECONDS));
+        timeCommand(event, removeCollectionNameFromCache(event), "SUCCESS", event.getElapsedTime(TimeUnit.NANOSECONDS));
     }
 
     @Override
     public void commandFailed(CommandFailedEvent event) {
-        timeCommand(event, "FAILED", event.getElapsedTime(TimeUnit.NANOSECONDS));
+        timeCommand(event, removeCollectionNameFromCache(event), "FAILED", event.getElapsedTime(TimeUnit.NANOSECONDS));
     }
 
-    private void timeCommand(CommandEvent event, String status, long elapsedTimeInNanoseconds) {
-        timerBuilder
+    private void timeCommand(
+            CommandEvent event,
+            String collection,
+            String status,
+            long elapsedTimeInNanoseconds
+    ) {
+        Timer.builder(TIMER_NAME)
+                .description(TIMER_DESC)
                 .tag("command", event.getCommandName())
                 .tag("cluster.id", event.getConnectionDescription().getConnectionId().getServerId().getClusterId().getValue())
                 .tag("server.address", event.getConnectionDescription().getServerAddress().toString())
+                .tag("collection", collection)
                 .tag("status", status)
                 .register(registry)
                 .record(elapsedTimeInNanoseconds, TimeUnit.NANOSECONDS);
     }
 
+    private void addCollectionNameToCache(CommandEvent event, String collectionName) {
+        if (collectionNamesCache.size() < collectionNamesCacheMaxSize) {
```

Comment:
A safety-net in case for some reason Mongo stops calling properly on command end and the cache is not getting cleaned (popped).